### PR TITLE
Adding more WireGuard FVs

### PR DESCRIPTION
**Pull request by realgaurav**
Adding more WireGuard FVs:
* WireGuard traffic test with GNP
* borrowed-ip handling in WireGuard route-table.
* nodes with mixed WireGuard enabled/disabled settings.

## Release Note
```release-note
None required
```

### CODE REVIEW

File: fv/wireguard_test.go [hunk 1]
```diff
@@ -351,6 +329,26 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
 			Expect(rcvd[1]).NotTo(BeEmpty())
 			Expect(rcvd[1]).To(Equal(sent[0]))
 		})
+
+		It("between pod to pod should be encrypted using wg tunnel with egress policy applied", func() {
```
Comment (robbrockbank) [inline fv/wireguard_test.go:333]: How do you know the policy is doing what you'd expect?

I think you should also have a similar test where the policy is set to "Deny" and you check the packet counts drop to 0.  Perhaps have it as part of the same test, so you modify the existing policy and then retry.

Comment (realgaurav) [reply fv/wireguard_test.go:341]: I had both tests initially but removed the `deny` test thinking `allow` test will be sufficient to invoke the policy code-path. Added both `allow` and `deny` tests now.

File: fv/wireguard_test.go [hunk 2]
```diff
@@ -348,6 +335,52 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
...
+					return tcpdumps[i].MatchCount("numInTunnelPackets")
+				}, "5s", "100ms").Should(BeNumerically("==", 0))
+				Consistently(func() int {
+					return tcpdumps[i].MatchCount("numOutTunnelPackets")
+				}, "5s", "100ms").Should(BeNumerically("==", 0))
+			}
+
+			cc.ResetExpectations()
+
+			policy.Name = "f01-egress-allow"
```
Comment (robbrockbank) [inline fv/wireguard_test.go:364]: I think this relies on the policy ordering being alphanumeric - which we don't actually expose anywhere.

I would either:
-  Modify the original policy, or
-  Set the Order number so this is definitely lower order than the first.

File: fv/wireguard_test.go [hunk 3]
```diff
@@ -348,6 +335,52 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
...
+			_, err = client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+
+			cc.ExpectSome(wls[0], wls[1])
+			cc.ExpectSome(wls[1], wls[0])
+			cc.CheckConnectivity()
+
+			By("verifying tunnelled packet count is non-zero")
+			for i := range felixes {
+				Eventually(func() int {
```
Comment (robbrockbank) [inline fv/wireguard_test.go:376]: For this to be useful it needs to check counts before and after you run connectivity.

IIUC, the first connectivity could result in tunnel packets while the policy is being installed.

Comment (robbrockbank) [reply fv/wireguard_test.go:376]: I've added a ResetCounts to the tcpdump - maybe useful once I've checked in my changes

File: fv/wireguard_test.go [hunk 4]
```diff
@@ -348,6 +335,52 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
...
+					return tcpdumps[i].MatchCount("numInTunnelPackets")
+				}, "10s", "100ms").Should(BeNumerically(">", 0))
+				Eventually(func() int {
+					return tcpdumps[i].MatchCount("numOutTunnelPackets")
+				}, "10s", "100ms").Should(BeNumerically(">", 0))
+			}
```
Comment (robbrockbank) [inline fv/wireguard_test.go:382]: I think 10s is gonna be too long for all of these tcp dump checks.  From a quick glance I think wireguard are the only tests that don't use the defaults (1s) for checking the tcp dump stats.  Do you think these need to be longer than the other tests?

Comment (realgaurav) [reply fv/wireguard_test.go:484]: There are a few tests using `tcpdump` in available examples. Original motivation to bump from `5s` to `10s` was flakes seen earlier, which may have been addressed by recent changes. In terms of impact though, I think polling interval will have actual impact for `Eventually` than the total time which is actually from `10ms` to `100ms` which isn't huge IMO. We can revisit this if needed.

File: fv/wireguard_test.go (hunk 2 above)
Comment (realgaurav) [reply fv/wireguard_test.go:466]: added order

File: fv/wireguard_test.go [hunk 5]
```diff
@@ -348,6 +335,52 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
...
+			policy.Spec.Egress = []api.Rule{{Action: api.Deny}}
+			policy.Spec.Selector = fmt.Sprintf("name in { '%s', '%s'}", wls[0].Name, wls[1].Name)
+			_, err := client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+
+			cc.ExpectNone(wls[0], wls[1])
+			cc.ExpectNone(wls[1], wls[0])
+			cc.CheckConnectivity()
+
+			By("verifying tunnelled packet count is non-zero")
```
[the remark this answers is not in the document]
Comment (realgaurav) [reply fv/wireguard_test.go:454]: done

File: fv/wireguard_test.go (hunk 3 above)
Comment (realgaurav) [reply fv/wireguard_test.go:482]: If I am understanding this comment correctly, we need to make sure that the policy is installed correctly before testing the connectivity. Will it be a good/clearer approach -- if we wait a few seconds and then read the policy and compare it with the policy created? If it matches then test the connectivity?

Comment (realgaurav) [reply fv/wireguard_test.go:504]: I updated the PR with suggested point above.

**Pull request merged**